### PR TITLE
3.0 [testing/api-client] Api tests for client <-> api interaction.

### PR DESCRIPTION
**Pull request by charlesg3**
**Please See** [Git Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions)

*Issue #, if available:*
PCLUSTER-2987

*Description of changes:*
Added an api-client fixture and a test that exercises it.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tests/integration-tests/conftest.py [hunk 1]
```diff
@@ -312,6 +313,60 @@ def _cluster_factory(cluster_config, extra_args=None, raise_on_error=True):
...
 
 
+@pytest.fixture(scope='class')
+def api_server_factory(cfn_stacks_factory, request, public_ecr_image_uri, api_definition_s3_uri):
+    """Creates a factory for deploying API servers on-demand to each region."""
+    api_servers = {}
+
+    def _api_server_factory(server_region):
+        api_stack_name = generate_stack_name("integ-tests-api", request.config.getoption("stackname_suffix"))
+        stack_template_path = os.path.join("..", "..", "api", "infrastructure", "parallelcluster-api.yaml")
```
Comment (demartinofra) [inline tests/integration-tests/conftest.py:323]: can be done separately but we want to use the template passed through the --api-infrastructure-s3-uri test-runner arg so that we decouple the artifacts under test from the checked-out version of integ-tests framework. This is particularly useful when testing released artifacts.

File: tests/integration-tests/test_runner.py [hunk 2]
```diff
@@ -305,6 +306,11 @@ def _init_argparser():
         help="S3 URI of the ParallelCluster API spec",
         default=TEST_DEFAULTS.get("public_ecr_image_uri"),
     )
+    debug_group.add_argument(
```
Comment (demartinofra) [inline tests/integration-tests/test_runner.py:309]: can we have an api_group and we move there all api related options?

File: tests/integration-tests/conftest.py [hunk 3]
```diff
@@ -312,6 +313,60 @@ def _cluster_factory(cluster_config, extra_args=None, raise_on_error=True):
         factory.destroy_all_clusters(test_passed=request.node.rep_call.passed)
 
 
+@pytest.fixture(scope='class')
+def api_server_factory(cfn_stacks_factory, request, public_ecr_image_uri, api_definition_s3_uri):
```
Comment (demartinofra) [inline tests/integration-tests/conftest.py:317]: are we sure the api_servers are reused across tests? isn't the class scoped fixture recreated and so the reference to the existing api_servers lost?

File: tests/integration-tests/conftest.py [hunk 4]
```diff
@@ -312,6 +313,60 @@ def _cluster_factory(cluster_config, extra_args=None, raise_on_error=True):
...
+        return api_servers[server_region]
+
+    yield _api_server_factory
+
+
+@pytest.fixture(scope="class")
+def api_client(region, api_server_factory, api_uri):
+    """Define a fixture for an API client that interacts with the
+    pcluster api."""
+    from pcluster_client.api import cluster_operations_api
```
Comment (demartinofra) [inline tests/integration-tests/conftest.py:355]: will the renaming be part of a separate PR?

File: tests/integration-tests/tests/api/test_api_client/test_api_client/pcluster.config.yaml [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+HeadNode:
+  InstanceType: {{ instance }}
+  Networking:
+    SubnetId: {{ public_subnet_id }}
+  Ssh:
+    KeyName: {{ key_name }}
+Scheduling:
+  Scheduler: {{ scheduler }}
+  SlurmSettings:
+    ScaledownIdletime: {{ scaledown_idletime }}
```
Comment (demartinofra) [inline tests/integration-tests/tests/api/test_api_client/test_api_client/pcluster.config.yaml:12]: feel free to remove this and keep the config minimal

File: tests/integration-tests/tests/api/test_api_client.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+from tests.common.utils import generate_random_string
+
+# Client Interface
+from pcluster_client.model.create_cluster_request_content import CreateClusterRequestContent
+
+@pytest.mark.regions(["us-east-2"])
+@pytest.mark.instances(["c5.xlarge"])
+@pytest.mark.schedulers(["slurm"])
+@pytest.mark.oss(["alinux2"])
```
Comment (demartinofra) [inline tests/integration-tests/tests/api/test_api_client.py:26]: feel free to remove these markers that are not required with the new test config approach. we will have to remove them from every test function sooner or later

File: tests/integration-tests/tests/api/test_api_client.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+
+def _test_describe_cluster(client, cluster, region):
+    response = client.describe_cluster(cluster.name, region=region)
+    assert_that(response['cluster_name']).is_equal_to(cluster.name)
+
+
+def _test_create(api_client, cluster_config, region):
+    cluster_config_data = base64.b64encode(cluster_config.encode('utf-8')).decode('utf-8')
+    body = CreateClusterRequestContent(f"integ-tests-{generate_random_string()}",
```
Comment (demartinofra) [inline tests/integration-tests/tests/api/test_api_client.py:54]: you could use the generate_stack_name from utils so that also the suffix is added to the name. This is used to clean-up stacks that are not properly deleted for some reason.
we could also transform generate_stack_name into a fixture so that the suffix is automatically injected from config options

File: tests/integration-tests/conftest.py [hunk 8]
```diff
@@ -312,6 +313,60 @@ def _cluster_factory(cluster_config, extra_args=None, raise_on_error=True):
...
+    if api_uri:
+        host = api_uri
+    else:
+        stack = api_server_factory(region)
+        host = stack.cfn_outputs["ParallelClusterApiInvokeUrl"]
+
+    api_configuration = Configuration(host=host)
+
+    with ApiClient(api_configuration) as api_client_instance:
+        yield cluster_operations_api.ClusterOperationsApi(api_client_instance)
```
Comment (demartinofra) [inline tests/integration-tests/conftest.py:367]: do you plan to return api_client_instance at some point to test endpoints that are not part of the ClusterOperationsApi?

File: tests/integration-tests/tests/api/test_api_client.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+# Client Interface
+from pcluster_client.model.create_cluster_request_content import CreateClusterRequestContent
+
+@pytest.mark.regions(["us-east-2"])
+@pytest.mark.instances(["c5.xlarge"])
+@pytest.mark.schedulers(["slurm"])
+@pytest.mark.oss(["alinux2"])
+@pytest.mark.usefixtures("region", "os", "instance")
+def test_api_client(scheduler, region, pcluster_config_reader, clusters_factory, api_client):
```
Comment (demartinofra) [inline tests/integration-tests/tests/api/test_api_client.py:28]: when adding tests for the custom image functionalities I'd put them in a separate test function so that image and cluster tests run in parallel.

File: tests/integration-tests/tests/api/test_api_client/test_api_client/pcluster.config.yaml [hunk 10]
```diff
@@ -0,0 +1,21 @@
+Image:
```
Comment (demartinofra) [inline tests/integration-tests/tests/api/test_api_client/test_api_client/pcluster.config.yaml:1]: as a reminder for a separate PR: it might be good to make this cluster config use as many features as possible so that we can better test the IAM policy attached to the API.

File: tests/integration-tests/conftest.py (hunk 1 above)
Comment (charlesg3) [reply tests/integration-tests/conftest.py:323]: This parameter didn't exist in my branch, so I added it and made it an override (so that things still work easily locally).

File: tests/integration-tests/test_runner.py (hunk 2 above)
Comment (charlesg3) [reply tests/integration-tests/test_runner.py:309]: :+1: done.

File: tests/integration-tests/conftest.py (hunk 4 above)
Comment (charlesg3) [reply tests/integration-tests/conftest.py:362]: I was waiting to hear if it's worth it to try to do a "namespace" package (which would involve moving files), but if not the change to the package name is here:

https://github.com/aws/aws-parallelcluster/pull/2899

File: tests/integration-tests/conftest.py (hunk 8 above)
Comment (charlesg3) [reply tests/integration-tests/conftest.py:374]: Yes --  I sent DM separately, but figured this could be an exercise of getting things end-to-end and expanded to provide the various clients (especially as the first round is iterated on). It's possible to:
- add the other clients here as well
- just yield the `api_client_instance` and then create the `cluster_operations_api` from within the test.

Comment (charlesg3) [reply tests/integration-tests/conftest.py]: Upon further thought, the second option feels better, at least for now so I pushed that change.

File: tests/integration-tests/tests/api/test_api_client.py (hunk 9 above)
Comment (charlesg3) [reply tests/integration-tests/tests/api/test_api_client.py]: Agreed -- I renamed this function to `test_cluster_operations` so that is more obvious.

File: tests/integration-tests/conftest.py (hunk 3 above)
Comment (charlesg3) [reply tests/integration-tests/conftest.py]: Converted to session based fixtures and tested by adding two tests to simply list clusters, the second test used a cached instance of the server (did not re-create the stack) then the stack was deleted at the end of the tests.

**Comment by demartinofra** [comment]
You can run the following to fix formatting and check code-linters:
```
cd cli
tox -e autoformat
tox -e code-linters
```

File: tests/integration-tests/conftest.py [hunk 11]
```diff
@@ -313,6 +317,62 @@ def _cluster_factory(cluster_config, extra_args=None, raise_on_error=True):
...
+        return api_servers[server_region]
+
+    yield _api_server_factory
+
+
+@pytest.fixture(scope="class")
+def api_client(region, api_server_factory, api_uri):
+    """Define a fixture for an API client that interacts with the
+    pcluster api."""
+    from pcluster_client import Configuration, ApiClient
```
Comment (demartinofra) [inline tests/integration-tests/conftest.py:362]: how is this import working?are you manually modifying your pythonpath to add the client dir?
I'm wondering if it makes sense to add the relative path to the client source in the requirements.txt defined under tests/integration-tests. The downside could be that `pip install -r requirements` will only work from within the tests/integration-tests dir.

Comment (charlesg3) [reply tests/integration-tests/conftest.py:362]: I did a `pip install -e .` on the generated code...

_something_ needs to happen but I wasn't sure if it's jenkins as a separate external step or embedded in req's (or something else)?

**Comment by ermanno** [comment]
I rebased the branch and added a commet with the checkstyle fixes.

**Review by demartinofra — APPROVED** [review]

**Comment by ermanno** [comment]
As agreed with @demartinofra I have squashed together the various commits, and removed the changes to `tests/integration-tests/configs/*` as the dev Jenkins is not yet ready to run these tests.

**Review by demartinofra — APPROVED** [review]
